posts/inside-rust/2020-10-30-exploring-pgo-for-the-rust-compiler.md

```diff
@@ -0,0 +1,299 @@
+But wall time measurements can be very noisy (which is why most people only look at instruction counts on perf.rust-lang.org), and `rustc-perf` only does a single iteration for each benchmark, so I was not prepared to trust these numbers just yet.
+I decided to try and reduce the noise by increasing the number of benchmark iterations from one to twenty.
+I only did "full" builds in this configuration as PGO's effect seemed to translate pretty predictably to incremental builds.
+After roughly eight hours to complete both the PGO and the non-PGO versions of the benchmarks these are the numbers I got:
+
+[![Wall time improvements gained from applying PGO to the entire compiler][rustc-perf-pgo-both-walltime-thump]][rustc-perf-pgo-both-walltime]
+
+[rustc-perf-pgo-both-walltime-thump]: /images/inside-rust/2020-10-30-exploring-pgo-for-the-rust-compiler/rustc-perf-pgo-both-walltime-thump.png
+[rustc-perf-pgo-both-walltime]: /images/inside-rust/2020-10-30-exploring-pgo-for-the-rust-compiler/rustc-perf-pgo-both-walltime.png
+
+As you can see we get a 10-16% reduction of build times almost across the board.
+This was more in line with what I had initially hoped to get from PGO.
+It is a bit surprising that the difference between instruction counts and wall time is so pronounced.
+One plausible explanation would be that PGO improves instruction cache utilization, something which makes a difference for execution time but would not be reflected in the amount of instructions executed.
+I also don't know how branch mispredictions factor into instruction counts -- branch prediction being another aspect explicitly targeted by PGO.
+
+As good as these numbers look, please keep in mind that they come from a single machine.
+It's possible that the Ryzen 1700X processor I used has some idiosyncrasies that favor the kind of optimizations that PGO does, and a different processor with a different caching system and branch predictor would generate quite different numbers.
+Nonetheless, the numbers undoubtedly are very encouraging and warrant further investigation.
+
+## Where to go from here
+
+The numbers above suggest that PGO can indeed provide noticeable compile time reductions.
+Unfortunately, bringing these improvements to end users is not as simple as adding a few compiler flags to our [dist builds][dist-builds].
+PGO is different from most other optimizations in that it
+
+ - requires a different, extended build workflow due to the additional instrumentation and data collection phases, and
+ - it incurs a sustained build time cost (a trait it shares with other automated optimizations like LTO).
+
+Both of these problems pose substantial hurdles for actually using PGO on the compiler itself.
+Rust's CI build times have always been too long and we already forgo some optimizations because of them
+(e.g. macOS still does not get the 10% performance boost from using a ThinLTOed LLVM because the build machines on that platform are especially slow).
```

Comment:
Do we currently use LTO on the compiler itself? I couldn't find anything in config.toml or any of the Cargo.toml files. That might be an easy start to this, running LTO on only the stable and beta channels without having to mess with the new PGO build system yet.

Reply:
LLVM (used to be) ThinLTOd on x86_64-unknown-linux-gnu, see https://github.com/rust-lang/rust/pull/78148 and linked issues/PRs for details. It wasn't a big loss to disable it (3-4% IIRC).

Reply:
Right - did we ever run LTO on the rust code though (as separate from LLVM)?

Reply:
Rust code I think is always ThinLTOd, though I might be misremembering. I don't recall if we've tried full LTO though. It should be relatively easy to try.

Reply:
> LLVM (used to be) ThinLTOd on x86_64-unknown-linux-gnu, see [rust-lang/rust#78148](https://github.com/rust-lang/rust/pull/78148) and linked issues/PRs for details. It wasn't a big loss to disable it (3-4% IIRC).

Interesting. I saw a bigger impact recently (~10% instruction count reduction, I think). I'll take another look.